cpp/daal/src/algorithms/kernel_function/polynomial/kernel_function_polynomial_dense_default_impl.i

```diff
@@ -0,0 +1,243 @@
+/* file: kernel_function_polynomial_dense_default_impl.i */
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __KERNEL_FUNCTION_POLYNOMIAL_DENSE_DEFAULT_IMPL_I__
+#define __KERNEL_FUNCTION_POLYNOMIAL_DENSE_DEFAULT_IMPL_I__
+
+#include "src/algorithms/kernel_function/polynomial/kernel_function_types_polynomial.h"
+
+#include "src/externals/service_blas.h"
+#include "src/externals/service_stat.h"
+#include "src/threading/threading.h"
+#include "src/algorithms/service_error_handling.h"
+#include "src/externals/service_math.h"
+
+namespace daal
+{
+namespace algorithms
+{
+namespace kernel_function
+{
+namespace polynomial
+{
+namespace internal
+{
+template <typename algorithmFPType, CpuType cpu>
+services::Status KernelImplPolynomial<defaultDense, algorithmFPType, cpu>::computeInternalVectorVector(const NumericTable * a1,
+                                                                                                       const NumericTable * a2, NumericTable * r,
+                                                                                                       const KernelParameter * par)
+{
+    if (par->kernelType != KernelType::linear)
+    {
+        return services::ErrorMethodNotImplemented;
+    }
+
+    //prepareData
+    const size_t nFeatures = a1->getNumberOfColumns();
+
+    ReadRows<algorithmFPType, cpu> mtA1(*const_cast<NumericTable *>(a1), par->rowIndexX, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtA1);
+    const algorithmFPType * dataA1 = mtA1.get();
+
+    ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), par->rowIndexY, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtA2);
+    const algorithmFPType * dataA2 = mtA2.get();
+
+    WriteOnlyRows<algorithmFPType, cpu> mtR(r, par->rowIndexResult, 1);
+    DAAL_CHECK_BLOCK_STATUS(mtR);
+    algorithmFPType * dataR = mtR.get();
+
+    //compute
+    dataR[0] = 0.0;
+    PRAGMA_IVDEP
+    PRAGMA_VECTOR_ALWAYS
+    for (size_t i = 0; i < nFeatures; i++)
+    {
+        dataR[0] += dataA1[i] * dataA2[i];
```

Comment:
Dot product?

Also, it should solve a potential issue that your implementation is not threaded.

Reply:
I would leave the implementation as a reference, because this is a code branch that is currently only for SMO method. Parallel by nFeatures? I'm not sure if this might be a good idea. :)

Reply:
MKL's dot product decides this internally, no need to do it manually.

Reply:
This part of the code is used only by linear kernel. I think it's out of scope of this PR and it's better to do it later

Reply:
1 line of code changes....

OK